### PR TITLE
Code generation with Jinja templates

### PR DESCRIPTION
**Pull request by equationcrunchor**
In addition to switching to Jinja,
- References to unimplemented `CANlib_Transmit_` functions were removed
- CANlib is no longer hardcoded to STM32F4 but now needs architecture to be specified at compile time
- Definitions of `CANlib_Identify` functions were added

### CODE REVIEW

File: generator/main.py [hunk 1]
```diff
@@ -1,37 +1,95 @@
...
+    constants.write(template_env, constants_template_path, constants_path)
+    render_template(template_env, "pack_unpack.c")
+    render_template(template_env, "pack_unpack.h")
+    render_template(template_env, "enum_atom.h")
+    render_template(template_env, "send_receive.c")
+    render_template(template_env, "structs.h")
+    render_template(template_env, "bus.h")
+    computers_h.write(template_env, system.computer, computer_h_template_path, computer_h_dir_path)
+    computers_c.write(template_env, system.computer, computer_c_template_path, computer_c_dir_path)
+    # test_h.write(can)
```
Comment (nistath) [inline generator/main.py:92]: Do you need these?

File: generator/main.py [hunk 2]
```diff
@@ -1,37 +1,95 @@
...
+        script_dir = '.'
+    os.chdir(script_dir)
+
+    template_loader = jinja2.FileSystemLoader(searchpath=".")
+    template_env = jinja2.Environment(loader=template_loader, keep_trailing_newline=True, extensions=[RaiseExtension])
+    template_env.globals["can"] = can
+    template_env.globals["system"] = system
+
+    constants.write(template_env, constants_template_path, constants_path)
+    render_template(template_env, "pack_unpack.c")
```
Comment (nistath) [inline generator/main.py:84]: Can you fold these into a for loop over a list of strings?

File: generator/main.py [hunk 3]
```diff
@@ -1,37 +1,95 @@
...
+    render_template(template_env, "pack_unpack.h")
+    render_template(template_env, "enum_atom.h")
+    render_template(template_env, "send_receive.c")
+    render_template(template_env, "structs.h")
+    render_template(template_env, "bus.h")
+    computers_h.write(template_env, system.computer, computer_h_template_path, computer_h_dir_path)
```
Comment (nistath) [inline generator/main.py:90]: Also fold these into a for loop?

File: generator/main.py [hunk 4]
```diff
@@ -1,37 +1,95 @@
...
-import enum_atom
-import structs
-import sys
 import computers_h
 import computers_c
-import test_h
-import test_c
-import send_receive
-import bus
+# import test_h
```
Comment (nistath) [inline generator/main.py:10]: Do you need these?

File: generator/main.py [hunk 5]
```diff
@@ -1,37 +1,95 @@
...
+# import test_c
 import drivers_inc
 
+from jinja2 import nodes
+from jinja2.ext import Extension
+from jinja2.exceptions import TemplateRuntimeError
+
+
+src_dir = '../src/'
```
Comment (nistath) [inline generator/main.py:19]: I would prefer pathlib be used for these. I see you have some instances of `{src_dir}/stuff` and some of `{src_dir}stuff`.

Comment (nistath) [reply generator/main.py:19]: `Path.join` would make this nicer and also work on other OSes.

File: generator/templates/computer.c.j2 [hunk 6]
```diff
@@ -0,0 +1,132 @@
+{%- macro handle_case(busnm, msg, msg_name=msg.name, indent=2) -%}
+{%- if msg.frame is defined -%}
+case CANlib_{{ busnm }}_{{ msg_name }}_key: {
```
Comment (nistath) [inline generator/templates/computer.c.j2:3]: What's the `:` at the end doing?

File: generator/templates/computer.c.j2 [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+  uint64_t bitstring;
+  to_bitstring(frame->data, &bitstring);
+  uint{{ (((msg.slice.length / 8) | round(method='ceil')) * 8) | int }}_t {{ busnm }}_{{ msg_name }}_key = EXTRACT(bitstring, {{ msg.slice.start }}, {{ msg.slice.length }});
```
Comment (nistath) [inline generator/templates/computer.c.j2:6]: Might be neater to have a macro or function for rounding up to a multiple of 8.

Comment (nistath) [reply generator/templates/computer.c.j2:6]: And reuse it where possible, such as line 28 below.

File: src/evil_macros.h [hunk 8]
```diff
@@ -1,51 +1,41 @@
...
     ((input) &= ~(1 << (7 - (idx)))); \
-  } \
+  }
 
-#define CHECK(a,b) (((a) & (1<<(7- (b)))) != 0)
+#define CHECK(a, b) (((a) & (1 << (7 - (b)))) != 0)
 
-#define ONES(len) \
-  ((1ULL << (len)) - 1)
+#define ONES(len) ((1ULL << (len)) - 1)
```
Comment (nistath) [inline src/evil_macros.h:16]: This doesn't work for 64 ones BTW. Probably need a special case for that.

File: src/evil_macros.h [hunk 9]
```diff
@@ -1,51 +1,41 @@
...
 
 #define INSERT(input, output, start, len) \
   (((output) & (ZEROES_MASK(start, len))) | INPUT_MASK(input, start, len))
 
 #define SIGN(input, data_width) \
-  ( \
-   ((int64_t)((input) << (64 - (data_width)))) >> \
-   (64 - (data_width)) \
-  )
+  (((int64_t)((uint64_t)(input) << (64 - (data_width)))) >> (64 - (data_width)))
```
Comment (nistath) [inline src/evil_macros.h:28]: I am not sure this is the right way to put the sign on. Could it be off by 1?

File: generator/main.py (hunk 1 above)
Comment (equationcrunchor) [reply generator/main.py:92]: No.

File: src/evil_macros.h (hunk 9 above)
Comment (equationcrunchor) [reply src/evil_macros.h:28]: I did some testing; it seems to be correct.

**Pull request merged**